src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -1901,33 +1901,31 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
             CollectionName collectionName,
             string docChangeVector,
             long lastModifiedTicks,
-            string changeVector,
+            ChangeVector changeVector,
             DocumentFlags flags,
             NonPersistentDocumentFlags nonPersistentFlags)
         {
             var newEtag = GenerateNextEtag();
 
-            if (string.IsNullOrEmpty(changeVector) &&
-                nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromReplication) == false)
+            if (nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromReplication))
             {
-                var mergedChangeVector = ConflictsStorage.GetMergedConflictChangeVectorsAndDeleteConflicts(
-                    context,
-                    lowerId,
-                    newEtag,
-                    context.GetChangeVector(docChangeVector).Version);
-
-                if (string.IsNullOrEmpty(mergedChangeVector))
-                    ChangeVectorUtils.ThrowConflictingEtag(lowerId.ToString(), docChangeVector, newEtag, Environment.Base64Id, DocumentDatabase.ServerStore.NodeTag);
-
-                context.LastDatabaseChangeVector ??= GetDatabaseChangeVector(context);
-                context.LastDatabaseChangeVector = ChangeVector.Merge(context.LastDatabaseChangeVector, context.GetChangeVector(mergedChangeVector), context);
-                changeVector = context.LastDatabaseChangeVector;
+                flags |= DocumentFlags.FromReplication;
             }
             else
             {
-                ConflictsStorage.DeleteConflictsFor(context, lowerId, null);
+                flags = flags.Strip(DocumentFlags.FromReplication);
+            }
+
+            var result = DocumentPut.BuildChangeVectorAndResolveConflicts(context, lowerId, newEtag, document: null, changeVector, expectedChangeVector: null, flags,
```

Comment:
something is off, if you had to make this public it means that this should not belong to DocumentPut? Move it to a ChangeVector or ChangeVectorHelper, or sth like that?